src/S3/Features/S3Transfer/MultipartDownloader.php

```diff
@@ -0,0 +1,281 @@
+
+    /**
+     * Calculates the object size dynamically.
+     *
+     * @param $sizeSource
+     *
+     * @return int
+     */
+    protected function computeObjectSize($sizeSource): int {
+        if (is_int($sizeSource)) {
+            return (int) $sizeSource;
+        }
+
+        if (empty($sizeSource)) {
+            throw new \RuntimeException('Range must not be empty');
+        }
+
+        // For extracting the object size from the ContentRange header value.
+        if (preg_match("/\/(\d+)$/", $sizeSource, $matches)) {
+            return $matches[1];
+        }
+
+        throw new \RuntimeException('Invalid source size format');
+    }
+
+    /**
+     * MultipartDownloader factory method to return an instance
+     * of MultipartDownloader based on the multipart download type.
+     *
+     * @param S3ClientInterface $s3Client
+     * @param string $multipartDownloadType
+     * @param array $requestArgs
+     * @param array $config
+     * @param int $currentPartNo
+     * @param int $objectPartsCount
+     * @param int $objectCompletedPartsCount
+     * @param int $objectSizeInBytes
+     * @param int $objectBytesTransferred
+     * @param string $eTag
+     * @param string $objectKey
+     * @param MultipartDownloadListener|null $listener
+     * @param TransferListener|null $progressTracker
+     *
+     * @return MultipartDownloader
+     */
+    public static function chooseDownloader(
+        S3ClientInterface $s3Client,
+        string $multipartDownloadType,
+        array $requestArgs,
+        array $config,
+        int $currentPartNo = 0,
+        int $objectPartsCount = 0,
+        int $objectCompletedPartsCount = 0,
+        int $objectSizeInBytes = 0,
+        int $objectBytesTransferred = 0,
+        string $eTag = "",
+        string $objectKey = "",
+        ?MultipartDownloadListener $listener = null,
+        ?TransferListener $progressTracker = null
+    ) : MultipartDownloader
+    {
+        return match ($multipartDownloadType) {
+            self::PART_GET_MULTIPART_DOWNLOADER => new GetMultipartDownloader(
+                s3Client: $s3Client,
+                requestArgs: $requestArgs,
+                config: $config,
+                currentPartNo: $currentPartNo,
+                objectPartsCount: $objectPartsCount,
+                objectCompletedPartsCount: $objectCompletedPartsCount,
+                objectSizeInBytes: $objectSizeInBytes,
+                objectBytesTransferred: $objectBytesTransferred,
+                eTag: $eTag,
+                objectKey: $objectKey,
+                listener: $listener,
+                progressListener: $progressTracker
+            ),
+            self::RANGE_GET_MULTIPART_DOWNLOADER => new RangeMultipartDownloader(
+                s3Client: $s3Client,
+                requestArgs: $requestArgs,
+                config: $config,
+                currentPartNo: 0,
+                objectPartsCount: 0,
```

Comment:
It seems like some of these things should be determined by the object itself rather than passed to it

Reply:
That is true, but, my intention is to also include a recoverable download functionality where customers can resume a download from a specific point, and in that scenario those values might need to be provided.